rotkehlchen/chain/evm/decoding/beefy_finance/decoder.py

```diff
@@ -0,0 +1,227 @@
+                asset_symbol = event.asset.resolve_to_asset_with_symbol().symbol
+
+                if is_withdrawal:
+                    event.event_subtype = HistoryEventSubType.RETURN_WRAPPED
+                    event.notes = f'Return {event.amount} {asset_symbol} to a Beefy vault'
+                else:
+                    event.event_type = HistoryEventType.DEPOSIT
+                    event.event_subtype = HistoryEventSubType.DEPOSIT_FOR_WRAPPED
+                    event.notes = f'Deposit {event.amount} {asset_symbol} in a Beefy vault'
+
+                spend_events.append(event)
+
+        # process receive events based on determined transaction type
+        for event in events:
+            if (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.event_subtype == HistoryEventSubType.NONE and
+                    (expected_amount is None or event.amount == expected_amount) and
+                    (expected_asset is None or event.asset == expected_asset)
+            ):
+                # Process the receive event
+                event.counterparty = CPT_BEEFY_FINANCE
+                asset_symbol = event.asset.resolve_to_asset_with_symbol().symbol
+
+                if is_withdrawal:
+                    event.event_type = HistoryEventType.WITHDRAWAL
+                    event.event_subtype = HistoryEventSubType.REDEEM_WRAPPED
+                    event.notes = f'Withdraw {event.amount} {asset_symbol} from a Beefy vault'
+                else:
+                    event.event_subtype = HistoryEventSubType.RECEIVE_WRAPPED
+                    event.notes = f'Receive {event.amount} {asset_symbol} after deposit in a Beefy vault'  # noqa: E501
+
+                receive_events.append(event)
+
+        return spend_events, receive_events
+
+    def _decode_deposit_or_withdrawal(
+            self,
+            transaction: EvmTransaction,  # pylint: disable=unused-argument
+            decoded_events: list['EvmEvent'],
+            all_logs: list[EvmTxReceiptLog],  # pylint: disable=unused-argument
+    ) -> list['EvmEvent']:
+        """Decodes vault interactions by identifying spend and receive events.
+
+        Transforms generic spend and receive events into structured deposit and withdrawal
+        operations for Beefy vaults. Deposits involve spending underlying tokens to receive
+        vault tokens, while withdrawals involve spending vault tokens to receive underlying assets.
+        """
+        spend_events, receive_events = self._process_beefy_events(decoded_events)
+        if len(spend_events) == 0 or len(receive_events) == 0:
+            log.error(f'Unable to find both spend and receive events for Beefy transaction {transaction}')  # noqa: E501
+            return decoded_events
+
+        maybe_reshuffle_events(
+            ordered_events=spend_events + receive_events,
+            events_list=decoded_events,
+        )
+        return decoded_events
```

Comment:
can't this check + maybe_reshuffle happen in `_process_beefy_events`. You are duplicating the code and seems that it belong to `_process_beefy_events`